### PR TITLE
Allow custom hashers and switch to abseil structures

### PR DESCRIPTION
**Pull request by dalloriam**
### Motivation
- Abseil containers generally perform better than standard containers
- Abseil containers support [heterogeneous lookup](https://abseil.io/tips/144), which will in turn allow us to support it in the cache 
  - (See https://github.com/coveooss/cachemere/pull/31)
### Overview
- Adds a dependency on abseil
- The main container of the cache changes from an `std::unordered_map` to `absl::node_hash_map`. `absl::flat_hash_map` would have been prefereable, but is not currently possible since cachemere's design requires pointer stability for the main container (since policies borrow keys & values from the main container)
- Associative containers used in policies can use `absl::flat_hash_map` (resulting in much faster lookups and less memory usage) since they don't need pointer stability.

### Motivation
- Implementations of `std::hash` (the current default) often are [horrendous hashing functions](https://martin.ankerl.com/2022/08/27/hashmap-bench-01/#stdhash-), especially for integral types.
- Implementations of `std::hash` cause gigantic performance problems with numeric keys and abseil containers.
- Hasher customization is a pre-requisite of heterogeneous lookup.
### Overview
- Added a new `KeyHash` template parameter to `cachemere::Cache` (and related cache presets)
- `KeyHash` has a default value of `absl::Hash<Key>`

-----
This is a large pull request, but all of the changes are pretty mechanical (the overwhelming majority of line changes are either replacing `std::hash` by `absl::Hash` _or_ adding the new `KeyHash` template parameter in _all_ cache and policy method signatures).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,77 +7,84 @@ option(USE_VCPKG "Use vcpkg for dependencies" OFF)
 if (USE_VCPKG)
     set(VCPKG_FEATURE_FLAGS "versions")
 
-    if(DEFINED ENV{VCPKG_ROOT})
+    if (DEFINED ENV{VCPKG_ROOT})
```
Comment (dalloriam) [inline CMakeLists.txt:10]: Note: these changes seem to be the output of clang-format formatting the CMake config. I don't have any objections to this (any automatically-enforced format is better than no format 🤷🏻 ) but I'm open to reverting the formatting if anyone has objections!

Comment (klalumiere) [reply CMakeLists.txt:10]: I'm usually not  too concerned with formatting, and this is not an exception. :slightly_smiling_face:

File: benchmarks/accuracy/src/io_benchmark.cpp [hunk 2]
```diff
@@ -29,7 +28,7 @@ struct Article {
     Article(const std::string& uri)
     {
         // Seed the RNG with a hash of our URI, this way items will have the same size & latency every time.
-        std::minstd_rand rng{static_cast<std::minstd_rand::result_type>(std::hash<std::string>()(uri))};
+        std::minstd_rand rng{static_cast<std::minstd_rand::result_type>(absl::Hash<std::string>()(uri))};
```
Comment (klalumiere) [inline benchmarks/accuracy/src/io_benchmark.cpp:32]: You should add the include related to this class (relying on transitive include is not robust and make the life or some tools harder).

**Review by klalumiere — COMMENTED** [review]
Just a small detail, and after I'll be down to merge.

**Review by klalumiere — APPROVED** [review]

**Pull request merged**